### PR TITLE
Promote to integration immediately given CI status

### PR DESCRIPTION
**Pull request by xbrianh**
* Add script to check combined CI status from GitHub
* Tweak `scripts/release.sh` to allow machine execution
* Promote to integration, at the end of the master build pipeline, when DCP-wide integration tests are passing

Note that the `--force` flag has been removed from `scripts/release.sh` in favor of the env var `DSS_FORCE_RELEASE`. By setting this env var in the GitLab UI, a pipeline can be retried with forced promotion if needed.

### CODE REVIEW

File: scripts/release.sh [hunk 1]
```diff
@@ -13,64 +13,70 @@ if [[ $# != 2 ]] && [[ $# != 3 ]]; then
     echo 'then this script runs "make deploy" after sourcing "environment.{DEST}",'
     echo "where DEST is the destination branch."
     echo
-    echo "If the --force flag is given, deployment will proceed even if CI checks fail."
+    echo "If ${DSS_FORCE_RELEASE}=='force', deployment will proceed even if CI checks fail."
```
Comment (kislyuk) [inline scripts/release.sh]: I am worried about the use of env vars to drive something as critical as this. Env vars are too implicit to toggle top level flags. I think it's preferable to call this script with `--force` in the caller's location.

File: scripts/release.sh [hunk 2]
```diff
@@ -13,64 +13,70 @@ if [[ $# != 2 ]] && [[ $# != 3 ]]; then
...
+    # Skip when this script is executed by the GitLab runner
+    echo "Please review and confirm your active AWS account configuration:"
+    aws configure list
+    aws sts get-caller-identity
+    aws iam list-account-aliases
+    echo "Is this correct?"
+    select result in Yes No; do
+        if [[ $result != Yes ]]; then exit 1; else break; fi
+    done
+fi
```
Comment (kislyuk) [inline scripts/release.sh]: I don't think this is a good pattern. It intertwines the concerns. What if I sudo to impersonate this user for some reason? Consider just calling this script with `yes |`:
```bash
yes Yes | scripts/release.sh
```

File: scripts/release.sh [hunk 3]
```diff
@@ -13,64 +13,70 @@ if [[ $# != 2 ]] && [[ $# != 3 ]]; then
...
+    
+    # TODO: (akislyuk) some CI builds no longer deploy or run a subset of tests. Find the last build that ran a deployment.
+    if [[ "$STATE" != success ]]; then
+        if [[ "force" == ${DSS_FORCE_RELEASE} ]]; then
+            echo "Status checks failed on branch $PROMOTE_FROM_BRANCH. Forcing promotion and deployment anyway."
+        else
+            echo "Status checks failed on branch $PROMOTE_FROM_BRANCH."
+            echo "Run with DSS_FORCE_RELEASE='force' to promote $PROMOTE_FROM_BRANCH to $PROMOTE_DEST_BRANCH and deploy anyway."
+            exit 1
+        fi
```
Comment (kislyuk) [inline scripts/release.sh]: Why are status checks skipped here on the CD runner?

File: scripts/status.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+GITHUB_API="https://api.github.com"
+
+OWNER=$(python -c "import sys, urllib.parse; print(urllib.parse.quote_plus(sys.argv[1]))" "${1}")
+REPO=$(python -c "import sys, urllib.parse; print(urllib.parse.quote_plus(sys.argv[1]))" "${2}")
+BRANCH=$(python -c "import sys, urllib.parse; print(urllib.parse.quote_plus(sys.argv[1]))" "${3}")
+
+STATUS=$(http GET ${GITHUB_API}/repos/${OWNER}/${REPO}/commits/${BRANCH}/status Accept:application/vnd.github.full+json)
+STATE=$(echo "$STATUS" | jq -r .state)
+
+echo ${STATE}
```
Comment (kislyuk) [inline scripts/status.sh]: Line 21 is unnecessary, you can just unwrap the subshell on line 19 if you don't need STATE for anything else.

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -73,3 +74,18 @@ deploy_dev:
...
+  stage: release_integration
+  script:
+    - while [[ "pending" == $(scripts/status.sh HumanCellAtlas dcp integration) ]]; do
+    -   echo "waiting for DCP Integration test to complete";
+    -   sleep 60;
+    - done
+    - if [[ "success" != $(scripts/status.sh HumanCellAtlas dcp integration) ]]; then
+    -   echo "DCP Integration test did not succeed";
+    -   exit 1
+    - fi
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply .gitlab-ci.yml]: The  dcp-wide integration test can be N commits ahead of data-store master. Which sha do we poll for?

Capped loop to 20 minutes.

File: Dockerfile.allspark [hunk 6]
```diff
@@ -42,21 +42,21 @@ RUN python3 -m pip install virtualenv==16.0.0
 RUN ln -s /usr/bin/python3.6 /usr/bin/python
 RUN ln -s /usr/bin/pip3 /usr/bin/pip
 
-RUN useradd -d /home/hca -ms /bin/bash -g root -G sudo hca
-RUN mkdir /HumanCellAtlas && chown hca /HumanCellAtlas
-USER hca
-WORKDIR /home/hca
+RUN useradd -d /home/hca_cicd -ms /bin/bash -g root -G sudo hca_cicd
+RUN mkdir /HumanCellAtlas && chown hca_cicd /HumanCellAtlas
+USER hca_cicd_cicd
```
[the remark this answers is not in the document]
Comment (xbrianh) [reply Dockerfile.allspark]: You can never have too many IMO.

fixed.

File: scripts/release.sh (hunk 1 above)
Comment (xbrianh) [reply scripts/release.sh]: That's reasonable, reverted to `--force`.

However, another mechanism will be needed if we wish to force promotion through the build box,

File: scripts/release.sh (hunk 2 above)
Comment (xbrianh) [reply scripts/release.sh]: Fixed.

File: scripts/release.sh (hunk 3 above)
Comment (xbrianh) [reply scripts/release.sh]: In order to reach this stage in the GitLab pipeline, testes have necessarily passed on the build box. Do we also wish to require  that Travis tests have passed?

File: scripts/status.sh (hunk 4 above)
Comment (xbrianh) [reply scripts/status.sh]: Fixed.

**Comment by xbrianh** [comment]
Perhaps the best place to document DSS ci/cd and release efforts is in a DSS-specific section for https://docs.google.com/document/d/1HbSWuIABaavaHpCwaO0peA_uwTW5wBCdRnYoz7ZVQuQ/edit#heading=h.yw6cvy5vlmbu

We can look into automating Dockerfile construction. In the meantime, added a pointer in `gitlab-ci.yml` to the usage instructions in `Dockerfile.allspark`.

File: .gitlab-ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (kislyuk) [reply .gitlab-ci.yml]: Can we poll for the sha of the commit from which this code is running? `git rev-parse HEAD`?

File: scripts/release.sh (hunk 3 above)
Comment (kislyuk) [reply scripts/release.sh]: I guess you're right, we don't want to depend on travis here.

**Comment by kislyuk** [comment]
Makes sense to document this in the SOP doc, but would be nice to move the SOP doc to the ops wiki now that discussion in google docs has completed.

File: .gitlab-ci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (kislyuk) [reply .gitlab-ci.yml]: Never mind... can't see an easy/elegant way to map the integration test repo sha to DSS sha for deployment OK.

**Pull request merged**